ts/protocol/utils.ts

```diff
@@ -192,3 +192,20 @@ export function timeoutPromise<T>(error: Error, timeoutMs: number): Promise<T> {
   let promiseResolver = new TimeoutPromiseResolver(error, timeoutMs);
   return promiseResolver.promise;
 }
+
+/**
+ * Is a cancellable PromiseResolver.
+ */
+export class CancellablePromise<T> extends PromiseResolver<T> {
+  constructor(private readonly value: T) {
+    super();
+  }
+
+  cancel() {
+    this.resolve(this.value);
```

Comment:
I think we don't reject `Promise`'s anywhere at the moment. Also adding result codes, will make this cleaner.